### ISSUE TITLE
How to set Firefox user_agent, proxy, profile and driver settings for the method instapy.set_selenium_remote_session?

### ISSUE THREAD

**Issue by mehmetaydar** [InstaPy/InstaPy#4922]
## Expected Behavior
In browser.py we have set_selenium_local_session method, which has mobile user agent, proxy, profile and other required properties set for Firefox. However the method set_selenium_remote_session in instapy class does not have them. Expected behavior is to have those in set_selenium_remote_session method.

## Current Behavior
It just does:
`self.browser = webdriver.Remote(
                command_executor=selenium_url,
                desired_capabilities=DesiredCapabilities.FIREFOX,
            )`

## Possible Solution (optional)
I am not sure if something like below would work:

       `capabilities = DesiredCapabilities.FIREFOX
            # set Firefox Agent to mobile agent
            user_agent = (
                "Mozilla/5.0 (iPhone; CPU iPhone OS 12_1 like Mac OS X) AppleWebKit/605.1.15 "
                "(KHTML, like Gecko) FxiOS/18.1 Mobile/16B92 Safari/605.1.15"
            )
            firefox_options = Firefox_Options()`

            `if headless_browser:
                firefox_options.add_argument("-headless")

            if browser_profile_path is not None:
                firefox_profile = webdriver.FirefoxProfile(browser_profile_path)
            else:
                firefox_profile = webdriver.FirefoxProfile()

            # set English language
            firefox_profile.set_preference("intl.accept_languages", "en-US")
            firefox_profile.set_preference("general.useragent.override", user_agent)

            if disable_image_load:
                # permissions.default.image = 2: Disable images load,
                # this setting can improve pageload & save bandwidth
                firefox_profile.set_preference("permissions.default.image", 2)

            if proxy_address and proxy_port:
                firefox_profile.set_preference("network.proxy.type", 1)
                firefox_profile.set_preference("network.proxy.http", proxy_address)
                firefox_profile.set_preference("network.proxy.http_port", proxy_port)
                firefox_profile.set_preference("network.proxy.ssl", proxy_address)
                firefox_profile.set_preference("network.proxy.ssl_port", proxy_port)

            # mute audio while watching stories
            firefox_profile.set_preference("media.volume_scale", "0.0")

            capabilities.setCapability(webdriver.Firefox.profile, firefox_profile)

            self.browser = webdriver.Remote(
                command_executor=selenium_url,
                desired_capabilities=capabilities
            )`

Also I am unable to figure out how to set gecko driver for webdriver.Remote.

**Comment by mehmetaydar** [comment]
Any help??

**Comment by ghost** [comment]
im waiting on this too

**Comment by mehmetaydar** [comment]
I partially made it work. However, I could not add instapy hide selenium extension to remote firefox on docker. Below is the code I am using:

> def set_selenium_remote_session(self, selenium_url='',
                                    selenium_driver=None,
                                    proxy_address=None,
                                    proxy_port=None,
                                    headless_browser=False,
                                    disable_image_load=False,
                                    browser_profile_path=None):
            capabilities = DesiredCapabilities.FIREFOX
            # set Firefox Agent to mobile agent
            user_agent = (
                "Mozilla/5.0 (iPhone; CPU iPhone OS 12_1 like Mac OS X) AppleWebKit/605.1.15 "
                "(KHTML, like Gecko) FxiOS/18.1 Mobile/16B92 Safari/605.1.15"
            )
            # keep user_agent
            Settings.user_agent = user_agent

            firefox_options = Firefox_Options()

            if headless_browser:
                firefox_options.add_argument("-headless")

            if browser_profile_path is not None:
                firefox_profile = webdriver.FirefoxProfile(browser_profile_path)
            else:
                firefox_profile = webdriver.FirefoxProfile()

            # set English language
            firefox_profile.set_preference("intl.accept_languages", "en-US")
            firefox_profile.set_preference("general.useragent.override", user_agent)

            if disable_image_load:
                # permissions.default.image = 2: Disable images load,
                # this setting can improve pageload & save bandwidth
                firefox_profile.set_preference("permissions.default.image", 2)

            if proxy_address and proxy_port:
                firefox_profile.set_preference("network.proxy.type", 1)
                firefox_profile.set_preference("network.proxy.http", proxy_address)
                firefox_profile.set_preference("network.proxy.http_port", proxy_port)
                firefox_profile.set_preference("network.proxy.ssl", proxy_address)
                firefox_profile.set_preference("network.proxy.ssl_port", proxy_port)

            # mute audio while watching stories
            firefox_profile.set_preference("media.volume_scale", "0.0")

	    """
             I tried to sign the hide selenium extension - I received an error
            """	
            firefox_profile.add_extension(firefox-ext/hide_selenium-1.0.0-an+fx.xpi")

            firefox_options._profile = firefox_profile
            capabilities.update(firefox_options.to_capabilities())

            self.browser = webdriver.Remote(
                #service_log_path=os.devnull,
                command_executor=selenium_url,
                desired_capabilities=capabilities
            )


            # authenticate with popup alert window
            #if proxy_username and proxy_password:
            #    proxy_authentication(browser, logger, proxy_username, proxy_password)

            self.browser.implicitly_wait(self.page_delay)

            # set mobile viewport (iPhone X)
            self.browser.set_window_size(375, 812)



        """
        #this was another way to add hide selenium extension, the url below was loading to the help page
        self.logger.info("Installing firefox extension")
        post_url = 'http://{0}-s:4444/session/{1}/moz/addon/install'.format(Settings.input_env, self.browser.session_id)
        self.logger.info("post_url: " + post_url)
        post_data = {
            "path": create_firefox_extension(), "temporary": True
        }
        r = requests.post(url=post_url, data = post_data ) 
        self.logger.info(r.text) """
        message = "Session started!"
        highlight_print("browser", message, "initialization", "info", self.logger)

        return self


Below is the log I am receiving from the firefox selenium container:

> test1-s | 18:24:33.471 INFO [ActiveSessionFactory.lambda$apply$11] - Matched factory org.openqa.selenium.grid.session.remote.ServicedSession$Factory (provider: org.openqa.selenium.firefox.GeckoDriverService)
test1-s | 1566930273626	mozrunner::runner	INFO	Running command: "/usr/bin/firefox" "-marionette" "-foreground" "-no-remote" "-profile" "/tmp/rust_mozprofile.8907IwqekwMc"
test1-s | 1566930273830	addons.xpi	WARN	Failed to parse recommendation: {"operation":"open","path":"/tmp/rust_mozprofile.8907IwqekwMc/extensions/instapy@rocks/mozilla-recommendation.json","unixErrno":2}
test1-s | 1566930273836	addons.xpi-utils	WARN	Add-on instapy@rocks is not correctly signed.
test1-s | 1566930273836	addons.xpi-utils	WARN	Add-on instapy@rocks is not correctly signed.
test1-s | 1566930273836	addons.xpi-utils	WARN	addMetadata: Add-on instapy@rocks is invalid: Error: Extension instapy@rocks is not correctly signed(resource://gre/modules/addons/XPIDatabase.jsm:2452:17) JS Stack trace: addMetadata@XPIDatabase.jsm:2452:17
test1-s | processFileChanges@XPIDatabase.jsm:2809:26
test1-s | checkForChanges@XPIProvider.jsm:2728:55
test1-s | startup@XPIProvider.jsm:2265:12
test1-s | callProvider@AddonManager.jsm:193:31
test1-s | _startProvider@AddonManager.jsm:569:5
test1-s | startup@AddonManager.jsm:725:14
test1-s | startup@AddonManager.jsm:2797:26
test1-s | observe@addonManager.js:65:29
test1-s | 1566930273849	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: mozillaAddons
test1-s | 1566930273849	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: telemetry
test1-s | 1566930273849	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: resource://pdf.js/
test1-s | 1566930273849	addons.webextension.screenshots@mozilla.org	WARN	Loading extension 'screenshots@mozilla.org': Reading manifest: Invalid extension permission: about:reader*
test1-s | 1566930274633	Marionette	INFO	Listening on port 44025
test1-s | 1566930274729	Marionette	WARN	TLS certificate errors will be ignored for this session
test1-s | 18:24:34.753 INFO [ProtocolHandshake.createSession] - Detected dialect: W3C
test1-s | 18:24:34.765 INFO [RemoteSession$Factory.lambda$performHandshake$0] - Started new session 280a166e-8a9a-407c-95f7-cd00f80002ae (org.openqa.selenium.firefox.GeckoDriverService)
test1-p | INFO [2019-08-27 18:24:34]   Session started!
test1-p | WARNING [2019-08-27 18:24:34]   Unable to update JSON state file
test1-p | INFO [2019-08-27 18:24:34]   -- Connection Checklist [1/3] (Internet Connection Status)
test1-p | INFO [2019-08-27 18:24:35]   - Internet Connection Status: ok
test1-p | INFO [2019-08-27 18:24:35]   - Current IP is "185.13.106.236" and it's from "United Kingdom/GB"
test1-p | WARNING [2019-08-27 18:24:35]   Unable to update JSON state file
test1-p | INFO [2019-08-27 18:24:35]   -- Connection Checklist [2/3] (Instagram Server Status)
test1-p | INFO [2019-08-27 18:24:51]   - Instagram WebSite Status: Currently Up 
test1-p | INFO [2019-08-27 18:24:51]   - Instagram Response Time: 40.526 ms 
test1-p | INFO [2019-08-27 18:24:51]   - Instagram Reponse Code: 200
test1-p | INFO [2019-08-27 18:24:51]   - Instagram Server Status: ok
test1-p | WARNING [2019-08-27 18:24:51]   Unable to update JSON state file
test1-p | INFO [2019-08-27 18:24:51]   -- Connection Checklist [3/3] (Hide Selenium Extension)
test1-p | INFO [2019-08-27 18:24:51]   - window.navigator.webdriver response: True
test1-p | WARNING [2019-08-27 18:24:51]   - Hide Selenium Extension: error
`
Any assistance would be useful. thanks.

**Comment by mehmetaydar** [comment]
I solved my issue by self signing the extension and using docker image of `selenium/standalone-firefox-debug:3.11`.

For self-signing, I copied arrive.js content.js and manifest.json into a folder called ext2.

 `cd ext2`

`npm install web-ext --global`

I created an account on firefox developer portal and generated signing keys.  

Under ext2 directory:
`web-ext sign --api-key myapiuser --api-secret my-api-secret --source-dir ext2"

`ls ext2/web-ext-artifacts/hide_selenium-1.0.0-an+fx.xpi`

In python code use: 

`        firefox_profile.add_extension(hide_selenium-1.0.0-an+fx.xpi")`
.
